yarn-project/simulator/src/avm/avm_tree.test.ts

```diff
@@ -0,0 +1,344 @@
+import { type BatchInsertionResult, type IndexedTreeId, MerkleTreeId, type MerkleTreeWriteOperations } from '@aztec/circuit-types';
+import {
+  NOTE_HASH_TREE_HEIGHT,
+  type NULLIFIER_TREE_HEIGHT,
+  type NullifierLeafPreimage,
+  type PUBLIC_DATA_TREE_HEIGHT,
+  PublicDataTreeLeaf,
+  type PublicDataTreeLeafPreimage,
+} from '@aztec/circuits.js';
+import { poseidon2Hash } from '@aztec/foundation/crypto';
+import { Fr } from '@aztec/foundation/fields';
+import { type IndexedTreeLeafPreimage } from '@aztec/foundation/trees';
+import { openTmpStore } from '@aztec/kv-store/utils';
+import { NoopTelemetryClient } from '@aztec/telemetry-client/noop';
+import { MerkleTrees } from '@aztec/world-state';
+
+import { AvmEphemeralForest, EphemeralAvmTree, type IndexedInsertionResult } from './avm_tree.js';
+
+let worldStateTrees: MerkleTrees;
+let copyState: MerkleTreeWriteOperations;
+// Up to 64 dummy note hashes
+let noteHashes: Fr[];
+let indexedHashes: Fr[];
+let slots: Fr[];
+let values: Fr[];
+let getSiblingIndex = 21n;
+
+// Helper to check the equality of the insertion results (low witness, insertion path)
+const checkEqualityOfInsertionResults = <Tree_Height extends number>(
+  containerResults: IndexedInsertionResult<IndexedTreeLeafPreimage>[],
+  wsResults: BatchInsertionResult<Tree_Height, 0>[],
+) => {
+  if (containerResults.length !== wsResults.length) {
+    throw new Error('Results length mismatch');
+  }
+  for (let i = 0; i < containerResults.length; i++) {
+    const containerResult = containerResults[i];
+    const wsResult = wsResults[i];
+    expect(containerResult.lowWitness.lowSiblingPath).toEqual(wsResult.lowLeavesWitnessData![0].siblingPath.toFields());
+    expect(containerResult.lowWitness.index).toEqual(wsResult.lowLeavesWitnessData![0].index);
+    expect(containerResult.lowWitness.preimage).toEqual(wsResult.lowLeavesWitnessData![0].leafPreimage);
+    expect(containerResult.insertionPath).toEqual(wsResult.newSubtreeSiblingPath.toFields());
+  }
+};
+
+beforeEach(async () => {
+  const store = openTmpStore(true);
+  worldStateTrees = await MerkleTrees.new(store, new NoopTelemetryClient());
+  copyState = await worldStateTrees.fork();
```

Comment:
why are these 2 different types? (MerkleTrees vs xxxOperations)

Reply:
Yeah I'm not sure (that wasnt me!) - i assume it's because whatever you get from `fork()` can't then be forked again 😃  hence this PR haha!